helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -643,14 +643,17 @@ public void init() {
     logger.info("initializing CallbackHandler: {}, content: {} ", _uid, getContent());
 
     if (_batchModeEnabled) {
-      synchronized (this) {
-        if (_batchCallbackProcessor != null) {
-          _batchCallbackProcessor.resetEventQueue();
+      CallbackProcessor callbackProcessorRef = _batchCallbackProcessor.get();
+        if (callbackProcessorRef != null) {
+          callbackProcessorRef.resetEventQueue();
         } else {
-          _batchCallbackProcessor = new CallbackProcessor(this);
-          _batchCallbackProcessor.start();
+          callbackProcessorRef = new CallbackProcessor(this);
+          if (_batchCallbackProcessor.compareAndSet(null, callbackProcessorRef)) {
+            callbackProcessorRef.start();
+          } else {
+            callbackProcessorRef.shutdown();
```

Comment:
I think `new CallbackProcessor(this)` creates a new thread and we need to close it if the new thread is not updated to `_batchCallbackProcessor`. However, since are doing start and then insert now, we need this `shutdown` anyway.